// gatsby-interface/dist/src/components/form/components/CheckboxFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { CheckboxField, CheckboxFieldWrapper, CheckboxFieldControl, CheckboxFieldLabel, CheckboxFieldHint, CheckboxFieldError } from './CheckboxField.esm.js';

var CheckboxFieldBlock = React__default.forwardRef(function CheckboxFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode"]);

  return jsx(CheckboxField, {
    id: id,
    hasError: !!error,
    hasHint: !!hint
  }, jsx(CheckboxFieldWrapper, {
    className: className
  }, jsx(CheckboxFieldControl, _extends({
    ref: ref
  }, rest)), jsx(CheckboxFieldLabel, {
    size: labelSize,
    isRequired: !!rest.required
  }, label), jsx(CheckboxFieldHint, null, hint), jsx(CheckboxFieldError, {
    validationMode: validationMode
  }, error)));
});
CheckboxFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/CheckboxFieldBlock.tsx"] = {
    name: "CheckboxFieldBlock",
    docgenInfo: CheckboxFieldBlock.__docgenInfo,
    path: "src/components/form/components/CheckboxFieldBlock.tsx"
  };
}

export { CheckboxFieldBlock };
