// gatsby-interface/dist/src/components/form/components/CheckboxGroupFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import { useFormFieldContainerProps } from './FormField.esm.js';
import { CheckboxGroupField, CheckboxGroupFieldLabel, CheckboxGroupFieldOptions, CheckboxGroupFieldOptionItem, CheckboxGroupFieldHint, CheckboxGroupFieldError } from './CheckboxGroupField.esm.js';

var CheckboxGroupFieldBlock = function CheckboxGroupFieldBlock(props) {
  var id = props.id,
      label = props.label,
      layout = props.layout,
      optionsDirection = props.optionsDirection,
      labelSize = props.labelSize,
      options = props.options,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      fieldValue = props.value,
      rest = _objectWithoutProperties(props, ["id", "label", "layout", "optionsDirection", "labelSize", "options", "error", "hint", "className", "validationMode", "value"]);

  var layoutProps = useFormFieldContainerProps(layout);
  return jsx(CheckboxGroupField, _extends({
    id: id,
    hasError: !!error,
    hasHint: !!hint,
    optionsDirection: optionsDirection,
    className: className
  }, layoutProps), jsx(CheckboxGroupFieldLabel, {
    size: labelSize,
    isRequired: !!rest.required,
    css: function css(_theme) {
      return [layout === "horizontal" && {
        alignSelf: "baseline"
      }];
    }
  }, label), jsx(CheckboxGroupFieldOptions, {
    css: function css(_theme) {
      return [layout === "horizontal" && {
        paddingTop: 0
      }];
    }
  }, options.map(function (_ref) {
    var value = _ref.value,
        label = _ref.label,
        restOption = _objectWithoutProperties(_ref, ["value", "label"]);

    return jsx(CheckboxGroupFieldOptionItem, _extends({
      key: value,
      value: value // Support uncontrolled field
      ,
      checked: fieldValue === undefined ? undefined : fieldValue.includes(value),
      label: label
    }, rest, restOption));
  })), jsx(CheckboxGroupFieldHint, null, hint), jsx(CheckboxGroupFieldError, {
    validationMode: validationMode
  }, error));
};
CheckboxGroupFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/CheckboxGroupFieldBlock.tsx"] = {
    name: "CheckboxGroupFieldBlock",
    docgenInfo: CheckboxGroupFieldBlock.__docgenInfo,
    path: "src/components/form/components/CheckboxGroupFieldBlock.tsx"
  };
}

export { CheckboxGroupFieldBlock };
