// gatsby-interface/dist/src/components/form/components/InputFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FormFieldContainer } from './FormField.esm.js';
import { InputField, InputFieldLabel, InputFieldControl, InputFieldHint, InputFieldError } from './InputField.esm.js';

var InputFieldBlock = React__default.forwardRef(function InputFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      layout = props.layout,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode", "layout"]);

  return jsx(FormFieldContainer, {
    layout: layout,
    className: className
  }, jsx(InputField, {
    id: id,
    hasError: !!error,
    hasHint: !!hint
  }, jsx(InputFieldLabel, {
    size: labelSize,
    isRequired: !!rest.required
  }, label), jsx(InputFieldControl, _extends({
    ref: ref
  }, rest)), jsx(InputFieldHint, null, hint), jsx(InputFieldError, {
    validationMode: validationMode
  }, error)));
});
InputFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputFieldBlock.tsx"] = {
    name: "InputFieldBlock",
    docgenInfo: InputFieldBlock.__docgenInfo,
    path: "src/components/form/components/InputFieldBlock.tsx"
  };
}

export { InputFieldBlock };
