// gatsby-interface/dist/src/components/form/components/SelectFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FormFieldContainer } from './FormField.esm.js';
import { SelectField, SelectFieldLabel, SelectFieldControl, SelectFieldHint, SelectFieldError } from './SelectField.esm.js';

var SelectFieldBlock = React__default.forwardRef(function SelectFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      layout = props.layout,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode", "layout"]);

  return jsx(FormFieldContainer, {
    layout: layout,
    className: className
  }, jsx(SelectField, {
    id: id,
    hasError: !!error,
    hasHint: !!hint
  }, jsx(SelectFieldLabel, {
    size: labelSize,
    isRequired: !!rest.required
  }, label), jsx(SelectFieldControl, _extends({
    ref: ref
  }, rest)), jsx(SelectFieldHint, null, hint), jsx(SelectFieldError, {
    validationMode: validationMode
  }, error)));
});
SelectFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectFieldBlock.tsx"] = {
    name: "SelectFieldBlock",
    docgenInfo: SelectFieldBlock.__docgenInfo,
    path: "src/components/form/components/SelectFieldBlock.tsx"
  };
}

export { SelectFieldBlock };
