// gatsby-interface/dist/src/components/form/components/TextAreaFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FormFieldContainer } from './FormField.esm.js';
import { TextAreaField, TextAreaFieldLabel, TextAreaFieldControl, TextAreaFieldHint, TextAreaFieldError } from './TextAreaField.esm.js';

var TextAreaFieldBlock = React__default.forwardRef(function TextAreaFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      layout = props.layout,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode", "layout"]);

  return jsx(FormFieldContainer, {
    layout: layout,
    className: className
  }, jsx(TextAreaField, {
    id: id,
    hasError: !!error,
    hasHint: !!hint
  }, jsx(TextAreaFieldLabel, {
    size: labelSize,
    isRequired: !!rest.required,
    css: function css(theme) {
      return [layout === "horizontal" && {
        alignSelf: "baseline",
        paddingTop: theme.space[2]
      }];
    }
  }, label), jsx(TextAreaFieldControl, _extends({
    ref: ref
  }, rest)), jsx(TextAreaFieldHint, null, hint), jsx(TextAreaFieldError, {
    validationMode: validationMode
  }, error)));
});
TextAreaFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/TextAreaFieldBlock.tsx"] = {
    name: "TextAreaFieldBlock",
    docgenInfo: TextAreaFieldBlock.__docgenInfo,
    path: "src/components/form/components/TextAreaFieldBlock.tsx"
  };
}

export { TextAreaFieldBlock };
